### PR TITLE
add AuthHeaderSigner

### PR DESCRIPTION
**Pull request by acharb**
[ticket](https://stellarorg.atlassian.net/browse/WAL-1390?atlOrigin=eyJpIjoiMWU4MTIwNzlhNjc1NDJjZjk5ODkwZjNmYzRjOWRlNmEiLCJwIjoiaiJ9)

[the kotlin change mimicking](https://github.com/stellar/kotlin-wallet-sdk/pull/128)

### CODE REVIEW

File: @stellar/typescript-wallet-sdk/src/walletSdk/Auth/index.ts [hunk 1]
```diff
@@ -165,3 +192,22 @@ const validateToken = (token: string) => {
...
+  account: AccountKeypair,
+  claims: AuthHeaderClaims,
+  clientDomain?: string,
+  authHeaderSigner?: AuthHeaderSigner,
+) => {
+  if (!authHeaderSigner) {
+    return null;
+  }
+
+  const issuer = clientDomain ? null : account;
```
Comment (acharb) [inline @stellar/typescript-wallet-sdk/src/walletSdk/Auth/index.ts:206]: @Ifropc just want to make sure this is right ([from kotlin](https://github.com/stellar/kotlin-wallet-sdk/blob/63b6e1bd2d93c2ce2b87cb4396d4025ffc87cf72/wallet-sdk/src/main/kotlin/org/stellar/walletsdk/auth/Sep10.kt#L211))

if the clientDomain is given, then issuer is null. That's right?

File: @stellar/typescript-wallet-sdk/src/walletSdk/Auth/AuthHeaderSigner.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  async createToken({
+    claims,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    clientDomain,
+    issuer,
+  }: AuthHeaderCreateTokenParams): Promise<string> {
+    if (!(issuer instanceof SigningKeypair)) {
+      throw new AuthHeaderSigningKeypairRequiredError();
+    }
+
```
Comment (acharb) [inline @stellar/typescript-wallet-sdk/src/walletSdk/Auth/AuthHeaderSigner.ts:54]: would love extra 👀 on this JWT creation code using nacl, to make sure it looks right

note: web api crypto [doesn't support ed25519](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/importKey#browser_compatibility) for most browsers, which is why I'm using nacl, which we're also using in the keymanager code

File: @stellar/typescript-wallet-sdk/src/walletSdk/Auth/AuthHeaderSigner.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    const naclKP = nacl.sign.keyPair.fromSeed(rawSeed);
+
+    // encode JWT message
+    const header = { alg: "EdDSA", typ: "JWT" };
+    const encodedHeader = base64url(JSON.stringify(header));
+    const encodedPayload = base64url(
+      JSON.stringify({ ...claims, exp: timeExp, iat: issuedAt }),
+    );
+
+    // sign JWT and create siganture
```
Comment (CassioMG) [inline @stellar/typescript-wallet-sdk/src/walletSdk/Auth/AuthHeaderSigner.ts:69]: "signature"

Comment (CassioMG) [inline @stellar/typescript-wallet-sdk/src/walletSdk/Auth/AuthHeaderSigner.ts:69]:
```suggestion
    // sign JWT and create signature
```

File: @stellar/typescript-wallet-sdk/src/walletSdk/Types/auth.ts [hunk 4]
```diff
@@ -91,3 +94,17 @@ export type SignChallengeTxnResponse = {
   transaction: XdrEncodedTransaction;
   networkPassphrase: NetworkPassphrase;
 };
+
+export type AuthHeaderClaims = {
+  account: string;
+  home_domain: string;
+  webAuthEndpoint: string;
```
Comment (CassioMG) [inline @stellar/typescript-wallet-sdk/src/walletSdk/Types/auth.ts:101]: should this be `web_auth_endpoint` instead?

Comment (acharb) [reply @stellar/typescript-wallet-sdk/src/walletSdk/Types/auth.ts]: yes, good 👁️  sir
https://github.com/stellar/typescript-wallet-sdk/pull/124/commits/84307862127c69aa96429093fb95b5a69fb910a3

File: @stellar/typescript-wallet-sdk/src/walletSdk/Auth/index.ts (hunk 1 above)
Comment (Ifropc) [reply @stellar/typescript-wallet-sdk/src/walletSdk/Auth/index.ts:206]: Yeah, for non-custodial wallets they need to set issuer equal to the domain account

File: @stellar/typescript-wallet-sdk/test/wallet.test.ts [hunk 5]
```diff
@@ -1880,3 +1884,56 @@ describe("Http client", () => {
     expect(resp.data.transaction).toBeTruthy();
   });
 });
+
+describe("AuthHeaderSigner", () => {
```
Comment (Ifropc) [inline @stellar/typescript-wallet-sdk/test/wallet.test.ts:1888]: IMO a good idea would be to test that this token works with anchor platform, but it's not merged yet (maybe once it's done)?
https://github.com/stellar/java-stellar-anchor-sdk/pull/1303

File: @stellar/typescript-wallet-sdk/test/wallet.test.ts [hunk 6]
```diff
@@ -1880,3 +1884,56 @@ describe("Http client", () => {
...
+  it("full sep-10 auth using header token should work", async () => {
+    const wallet = Wallet.TestNet();
```
Comment (Ifropc) [inline @stellar/typescript-wallet-sdk/test/wallet.test.ts:1893]: You can also check that JWT from SEP-10 matches JWT issued by TS SDK
https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md
(You can find private key in kotlin tests)

File: @stellar/typescript-wallet-sdk/test/wallet.test.ts (hunk 5 above)
Comment (acharb) [reply @stellar/typescript-wallet-sdk/test/wallet.test.ts:1888]: ya agreed, I'll add a small ticket to track

https://stellarorg.atlassian.net/browse/WAL-1441

File: @stellar/typescript-wallet-sdk/test/wallet.test.ts (hunk 6 above)
Comment (acharb) [reply @stellar/typescript-wallet-sdk/test/wallet.test.ts:1893]: err not sure what you mean here, could you explain further?

Comment (Ifropc) [reply @stellar/typescript-wallet-sdk/test/wallet.test.ts:1893]: I think you should run your code with the exact same parameters as in protocol examples and verify that token matches

Comment (acharb) [reply @stellar/typescript-wallet-sdk/test/wallet.test.ts:1893]: 👍  added a test to check against the example token, and removed the `typ` header

https://github.com/stellar/typescript-wallet-sdk/pull/124/commits/ad7c0d2534ddb8e7b09f8df9c1d6db062b1e18a8

**Comment by Ifropc** [comment]
LGTM!

File: @stellar/typescript-wallet-sdk/test/wallet.test.ts (hunk 6 above)
Comment (Ifropc) [reply @stellar/typescript-wallet-sdk/test/wallet.test.ts:1893]: Do you mind adding a test for custodial wallet too?

**Review by Ifropc — APPROVED** [review]

**Pull request merged**